File: Data Structures & Algorithms/minimum-array-end/submission-5.cpp
class Solution {
public:
    long long minEnd(int n, int x) {
        long long res = x;
        --n;        // we need to calculate for n-1
        int i = 0;
        
        while (n > 0) {
            if ((res&(1LL<<i)) == 0) { // if ith bit is 0 in x
                if (n&1) {       // if n's 0th bit is set, add it to x at ith position
                    res |= (1LL<<i);
                }
                n >>= 1;        // divide n by 2 to get rid of LSB
            }
            ++i;                // let's move to next bit of x
        }
        
        return res;
    }
};

File: Data Structures & Algorithms/valid-parenthesis-string/submission-5.java
class Solution {
    public boolean checkValidString(String s) {
        int n = s.length();
        boolean[][] dp = new boolean[n+1][n+1];
        dp[n][0] = true;

        for (int i = n-1; i >= 0; i--) {
            for (int open = 0; open < n; open++) {
                boolean res = false;

                if (s.charAt(i) == '(') {
                    res |= dp[i+1][open+1];
                } else if (s.charAt(i) == ')') {
                    if (open > 0) {
                        res |= dp[i+1][open-1];
                    }
                } else {
                    res |= dp[i+1][open+1];
                    
                    if (open > 0) {
                        res |= dp[i+1][open-1];
                    }

                    res |= dp[i+1][open];
                }

                dp[i][open] = res;
            }
        }

        return dp[0][0];
    }
}
